src/tactic/interval_cases.lean

```diff
@@ -0,0 +1,220 @@
+meta def combine_lower_bounds : option expr → option expr → tactic (option expr)
+| none none := return $ none
+| (some prf) none := return $ some prf
+| none (some prf) := return $ some prf
+| (some prf₁) (some prf₂) :=
+  do option.some <$> to_expr ``(max_le %%prf₁ %%prf₂)
+
+/-- Inspect a given expression, using it to update a set of upper and lower bounds on `n`. -/
+meta def update_bounds (n : expr) (bounds : option expr × option expr) (e : expr) :
+  tactic (option expr × option expr) :=
+do nlb ← try_core $ gives_lower_bound n e,
+   nub ← try_core $ gives_upper_bound n e,
+   clb ← combine_lower_bounds bounds.1 nlb,
+   cub ← combine_upper_bounds bounds.2 nub,
+   return (clb, cub)
+
+/--
+Attempt to find a lower bound for the variable `n`, by evaluating `lattice.bot_le n`.
+-/
+meta def initial_lower_bound (n : expr) : tactic expr :=
+do e ← to_expr ``(@lattice.bot_le _ _ %%n),
+   t ← infer_type e,
+   match t with
+   | `(%%b ≤ %%n) := do return e
+   | _ := failed
+   end
+
+/--
+Attempt to find an upper bound for the variable `n`, by evaluating `lattice.le_top n`.
+-/
+meta def initial_upper_bound (n : expr) : tactic expr :=
+do e ← to_expr ``(@lattice.le_top _ _ %%n),
+   match e with
+   | `(%%n ≤ %%b) := do
+     tn ← infer_type n,
+     e ← match tn with
+     | `(ℕ) := to_expr ``(nat.add_one_le_iff.mpr %%e)
+     | `(ℕ+) := to_expr ``(pnat.add_one_le_iff.mpr %%e)
+     | `(ℤ) := to_expr ``(int.add_one_le_iff.mpr %%e)
+     | _ := failed
+     end,
+     return e
+   | _ := failed
+   end
+
+/-- Inspect the local hypotheses for upper and lower bounds on a variable `n`. -/
+meta def get_bounds (n : expr) : tactic (expr × expr) :=
+do
+  hl ← try_core (initial_lower_bound n),
+  hu ← try_core (initial_upper_bound n),
+  lc ← local_context,
+  r ← lc.mfoldl (update_bounds n) (hl, hu),
+  match r with
+  | (_, none) := fail "No upper bound located."
+  | (none, _) := fail "No lower bound located."
+  | (some lb_prf, some ub_prf) := return (lb_prf, ub_prf)
+  end
+
+/-- The finset of elements of a set `s` for which we have `fintype s`. -/
+def set_elems {α} [decidable_eq α] (s : set α) [fintype s] : finset α :=
+(fintype.elems s).image subtype.val
+
+/-- Each element of `s` is a member of `set_elems s`. -/
+lemma mem_set_elems {α} [decidable_eq α] (s : set α) [fintype s] {a : α} (h : a ∈ s) :
+  a ∈ set_elems s :=
+finset.mem_image.2 ⟨⟨a, h⟩, fintype.complete _, rfl⟩
+
+end interval_cases
+
+open interval_cases
+
+/-- Call `fin_cases` on membership of the finset built from
+an `Ico` interval corresponding to a lower and an upper bound.
+
+Here `hl` should be an expression of the form `a ≤ n`, for some explicit `a`, and
+`hu` should be of the form `n < b`, for some explicit `b`.
+-/
+meta def interval_cases_using (hl hu : expr) : tactic unit :=
+to_expr ``(mem_set_elems (Ico _ _) ⟨%%hl, %%hu⟩) >>= note_anon >>= fin_cases_at none
+
+setup_tactic_parser
+
+namespace interactive
+
+/--
+`interval_cases hl hu` takes two hypotheses,
```

Comment:
```suggestion
`interval_cases_using hl hu` takes two hypotheses,
```

Reply:
Ugh, I really should just add a parser keyword for `using` to the main interactive tactic.

Reply:
Okay, I've combined the two interactive tactics, so now you either call
* `interval_cases n`, which goes looking for bounds itself, or
* `interval_cases using lb ub`, if you want to explicitly specify the lower and upper bounds.